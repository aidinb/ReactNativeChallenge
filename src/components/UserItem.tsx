import React from 'react';
import { TouchableOpacity, Text, StyleSheet, Dimensions } from 'react-native';
import { Item } from '../utils/types.ts';

interface UserItemProps {
  item: Item;
  onPress: (item: Item) => void;
}

const { width } = Dimensions.get('window');

const UserItem: React.FC<UserItemProps> = ({ item, onPress }) => (
    <TouchableOpacity
        onPress={() => onPress(item)}
        style={styles.card}
    >
      <Text style={styles.text}>{item.name}</Text>
    </TouchableOpacity>
);

const styles = StyleSheet.create({
  card: {
    width: width * 0.9,
    paddingVertical: 15,
    backgroundColor: '#ffffff',
    marginVertical: 10,
    borderRadius: 8,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.2,
    shadowRadius: 4,
    elevation: 2,
    alignSelf: 'center',
  },
  text: {
    fontSize: 16,
    color: '#333',
    paddingLeft: 15,
  },
});

export default React.memo(UserItem);
