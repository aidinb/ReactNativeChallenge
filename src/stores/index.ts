import React from 'react';
import MainStore from './mainStore.ts';

class RootStore {
  mainStore: MainStore;

  constructor() {
    this.mainStore = new MainStore();
  }
}

const storesContext = React.createContext<RootStore>(new RootStore());

// This will be the function available for the app to connect to the stores
export const useStores = (): RootStore => React.useContext(storesContext);
