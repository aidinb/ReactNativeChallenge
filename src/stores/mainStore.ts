import { action, makeObservable, observable } from 'mobx';
import { getItems } from '../utils/api';
import {Item} from '../utils/types.ts';

class MainStore {
  allUsers: Item[] = [];
  user: Item = {} as Item; // Use the Item type
  loading: boolean = false;
  error: string | null = null;

  constructor() {
    makeObservable(this, {
      allUsers: observable,
      user: observable,
      loading: observable,
      error: observable,
      getAllUsers: action,
      setAllUsers: action,
      setActiveUser: action,
      setLoading: action,
      setError: action,
    });
  }

  setAllUsers = (categories: Item[]) => {
    this.allUsers = categories;
  };

  setActiveUser = (user: Item) => {
    this.user = user;
  };

  setLoading = (value: boolean) => {
    this.loading = value;
  };

  setError = (message: string | null) => {
    this.error = message;
  };

  getAllUsers = async () => {
    this.setLoading(true);
    this.setError(null);
    try {
      const users = await getItems();
      this.setAllUsers(users);
    } catch (error) {
      this.setError('Failed to fetch users');
    } finally {
      this.setLoading(false);
    }
  };
}

export default MainStore;
