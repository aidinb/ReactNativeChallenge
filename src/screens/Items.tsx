import React, { useCallback, useEffect } from 'react';
import { View, StyleSheet, FlatList, ListRenderItem } from 'react-native';
import { observer } from 'mobx-react';
import { COLORS } from '../styles';
import NavigationService from '../navigation/NavigationService';
import { useStores } from '../stores';
import UserItem from '../components/UserItem';
import Loading from '../components/Loading';
import ErrorMessage from '../components/ErrorMessage';
import {Item} from '../utils/types.ts';

const Items: React.FC = () => {
    const { mainStore } = useStores(); // Access the store

    useEffect(() => {
        mainStore.getAllUsers();
    }, [mainStore]);

    const renderItems: ListRenderItem<Item> = useCallback(
        ({ item }) => (
            <UserItem
                item={item}
                onPress={() => {
                    mainStore.setActiveUser(item);
                    NavigationService.navigate('ItemDetail');
                }}
            />
        ),
        [mainStore]
    );

    return (
        <View style={styles.container}>
            {!mainStore.loading && !mainStore.error && (
                <FlatList
                    showsVerticalScrollIndicator={false}
                    keyExtractor={(item) => item.id.toString()} // Use id directly as string
                    data={mainStore.allUsers}
                    renderItem={renderItems}
                />
            )}
            {mainStore.loading && <Loading />}
            {mainStore.error && <ErrorMessage message={mainStore.error} />}
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        backgroundColor: COLORS.white,
        flex: 1,
    },
});

export default observer(Items);
