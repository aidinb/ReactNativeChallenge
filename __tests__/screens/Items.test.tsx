import React from 'react';
import { render, waitFor } from '@testing-library/react-native';
import { act } from 'react-test-renderer'; // Import act to handle async operations
import { useStores } from '../stores'; // Mock the useStores hook
import NavigationService from '../navigation/NavigationService.ts'; // Mock NavigationService
import UserItem from '../components/UserItem.tsx';
import Items from "../../src/screens/Items.tsx"; // Mock UserItem component

// Mock the useStores hook to control its output
jest.mock('../stores', () => ({
    useStores: () => ({
        mainStore: {
            getAllUsers: jest.fn(),
            allUsers: [{ id: '1', name: 'User 1' }, { id: '2', name: 'User 2' }],
            setActiveUser: jest.fn(),
        },
    }),
}));

// Mock NavigationService to prevent actual navigation
jest.mock('../navigation/NavigationService.ts', () => ({
    navigate: jest.fn(),
}));

// Mock UserItem component
jest.mock('../components/UserItem.tsx', () => 'UserItem');

describe('Items Screen', () => {
    it('renders correctly and displays user items', async () => {
        const { toJSON, getByText } = render(<Items />);

        // Check if the component renders correctly
        expect(toJSON()).toMatchSnapshot();

        // Check if user items are rendered
        await waitFor(() => {
            expect(getByText('User 1')).toBeTruthy();
            expect(getByText('User 2')).toBeTruthy();
        });
    });

    it('navigates to ItemDetail when a user item is pressed', async () => {
        const { getByText } = render(<Items />);

        // Simulate pressing a user item
        await act(async () => {
            // Find the first user item and simulate a press
            const userItem = getByText('User 1').closest('TouchableOpacity');
            userItem?.props.onPress();
        });

        // Verify that the navigation function was called
        expect(NavigationService.navigate).toHaveBeenCalledWith('ItemDetail');
    });
});
