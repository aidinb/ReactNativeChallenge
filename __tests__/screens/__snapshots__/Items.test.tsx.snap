// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items Screen renders correctly and displays user items 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={
      [
        {
          "id": "1",
          "name": "User 1",
        },
        {
          "id": "2",
          "name": "User 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#ffffff",
              "borderRadius": 8,
              "elevation": 2,
              "marginVertical": 10,
              "opacity": 1,
              "paddingVertical": 15,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.2,
              "shadowRadius": 4,
              "width": 675,
            }
          }
        >
          <Text
            style={
              {
                "color": "#333",
                "fontSize": 16,
                "paddingLeft": 15,
              }
            }
          >
            User 1
          </Text>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#ffffff",
              "borderRadius": 8,
              "elevation": 2,
              "marginVertical": 10,
              "opacity": 1,
              "paddingVertical": 15,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.2,
              "shadowRadius": 4,
              "width": 675,
            }
          }
        >
          <Text
            style={
              {
                "color": "#333",
                "fontSize": 16,
                "paddingLeft": 15,
              }
            }
          >
            User 2
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
